sdk/javascript/generator/StructTypeFormatter.py

```diff
@@ -0,0 +1,366 @@
+					if bound_condition:
+						condition_value = bound_field.value.value
+						field_value = f'({bound_field_name} ? {field_value} : {condition_value})'
+				else:
+					field_value = bound_field.extensions.printer.get_size()
+			elif field.is_size_reference:
+				field_value = bound_field.extensions.printer.get_size()
+		else:
+			field_value = self.field_name(field)
+
+		serialize_line = None
+		if DisplayType.TYPED_ARRAY == field.display_type:
+			serialize_field = field.extensions.printer.store(field_value, 'buffer_')
+			serialize_line = f'{serialize_field};{field_comment}\n'
+		else:
+			serialize_field = field.extensions.printer.store(field_value)
+			serialize_line = f'buffer_.write({serialize_field});{field_comment}\n'
+
+		return indent_if_conditional(condition, serialize_line)
+
+	def get_serialize_descriptor(self):
+		body = 'const buffer_ = new Writer(this.size);\n'
+
+		# if first field is size replace serializer with custom one (to access builder .size() instead)
+		fields_iter = self.non_const_fields()
+		first_field = next(fields_iter)
+		if self.struct.size == first_field.extensions.printer.name:
+			body += f'buffer_.write(converter.intToBytes(this.size, {first_field.size}, false));\n'
+		else:
+			body += self.generate_serialize_field(first_field)
+
+		for field in fields_iter:
+			body += self.generate_serialize_field(field)
+
+		body += 'return buffer_.storage;'
+		return MethodDescriptor(body=body)
+
+	def generate_size_field(self, field):
+		condition = self.generate_condition(field, True)
+		size_field = field.extensions.printer.get_size()
+
+		return indent_if_conditional(condition, f'size += {size_field};\n')
+
+	def get_size_descriptor(self):
+		body = 'let size = 0;\n'
+		body += ''.join(map(self.generate_size_field, self.non_const_fields()))
+		body += 'return size;'
+		return MethodDescriptor(body=body)
+
+	def create_getter_descriptor(self, field):
+		method_descriptor = MethodDescriptor(
+			method_name=f'get {field.extensions.printer.name}',
+			body=f'return {self.field_name(field)};',
+		)
+		return method_descriptor
+
+	def get_getter_descriptors(self):
+		return list(map(self.create_getter_descriptor, self.non_reserved_fields()))
+
+	def create_setter_descriptor(self, field):
+		method_descriptor = MethodDescriptor(
+			method_name=f'set {field.extensions.printer.name}',
+			arguments=['value'],
+			body=f'{self.field_name(field)} = value;',
+		)
+		return method_descriptor
+
+	def get_setter_descriptors(self):
+		return list(map(self.create_setter_descriptor, self.non_reserved_fields()))
+
+	def generate_str_field(self, field):
+		condition = self.generate_condition(field, True)
+		field_to_string = field.extensions.printer.to_string(self.field_name(field))
+		value = None
+		if DisplayType.TYPED_ARRAY == field.display_type:
+			value = f'[${{{field_to_string}}}]'
+		elif DisplayType.BYTE_ARRAY == field.display_type:
+			value = f'hex(${{{field_to_string}}})'
+		else:
+			value = f'${{{field_to_string}}}'
```

Comment:
hack to make toString formatting bit nicer, maybe should be moved down to printers.to_string